Integrate a Stellar asset contracts (#754)
stellar/stellar-docs: docs/build/guides/conventions/stellar-asset-contract.mdx

---
title: Integrate Stellar Assets Contracts
description: Test and use Stellar assets in a Soroban smart contract
---

> From the contract perspective Stellar Asset Contract is not different from any other token that implements the Soroban [SEP-41 Token Interface]. The Rust SDK contains a pregenerated client for any contract that implements the token interface:

Reviewer: ```suggestion
When interacting with assets in a smart contract, the Stellar Asset Contract is not different from any other token that implements the Stellar [SEP-41 Token Interface]. The Rust SDK contains a pre-generated client for any contract that implements the token interface:
```

Or something similar

```rust
use soroban_sdk::token;

> struct MyContract;

Reviewer: ```suggestion
#[contract]
struct MyContract;
```

I think we need the `#[contract]` macro here, right?

#[contractimpl]
impl MyContract {

>   fn token_fn(e: Env, id: Address) {

Reviewer: 4-space indentations in the Rust code examples, please. If you run `yarn format:mdx` in your local repo it "should" fix that for you (assuming you're editing locally, and have installed all the dependencies, that is).

Reviewer: Thanks for fixing this manually. There may be something weird happening with our formatter (probably a weird thing with the v3 upgrade). I'll do some digging 🙏🏻

    // Create a client instance for the provided token identifier. If the id
    // value corresponds to an SAC contract, then SAC implementation is used.
    let client = token::TokenClient::new(&env, &id);
    // Call token operations part of the SEP-41 token interface
    client.transfer(...);
  }
}
```

> The address parameter is not the address of the issuer of an asset, it corresponds to the deployed contract address for this asset.

Reviewer: ```suggestion
The `asset` parameter is not the address of the issuer of an asset, it corresponds to the deployed contract address for this asset.
```

```bash
stellar contract id asset \
    --source G... \
    --network testnet \
    --asset [asset:issuer]
```

E.g. for USDC, it would be `--asset USDC:G...` For the native asset, XLM, `--asset native`. See the [deploy SAC] guide for more details.

[deploy SAC]: ../cli/deploy-stellar-asset-contract.mdx

> :::info Clients

Reviewer: ```suggestion
:::info[Clients]
```

As part of the upgrade from docusaurus v2 to v3, the way we do [admonition titles changes](https://docusaurus.io/docs/migration/v3#admonitions-option). There is a compatibility flag that is turned on by default, but we'll want to migrate all existing admonition titles to use the new syntax. So, for new ones, let's use the new way to begin.

Author: A great thanks!

> A client created by [`token::Client`] implements the functions defined by any contract that implements the [SEP-41 Token Interface]. But with [CAP-46-6 smart contract standardized asset], the Stellar Asset Contract exposes additional functions such as `mint`. To access the additional functions, another client needs to be used: [`token::StellarAssetClient`]. This client only implements the functions from CAP-46-6, which are not part of the SEP-41.

Reviewer: ```suggestion
A client created by [`token::TokenClient`] implements the functions defined by any contract that implements the [SEP-41 Token Interface]. But with [CAP-46-6 smart contract standardized asset], the Stellar Asset Contract exposes additional functions such as `mint`. To access the additional functions, another client needs to be used: [`token::StellarAssetClient`]. This client only implements the functions from CAP-46-6, which are not part of the SEP-41 interface.
```

I think this link definition might be missing from this file, too.

```rust
let client = token::StellarAssetClient::new(&env, &id);
// Call token operations which are not part of the SEP-41 token interface
// but part of the CAP-46-6 Smart Contract Standardized Asset
client.mint(...);
```

:::

### Testing

Soroban Rust SDK provides an easy way to instantiate a Stellar Asset Contract tokens using `register_stellar_asset_contract`. This function can be seen as the deployment of a generic token. In the following example, we are following the best practices outlined in the [Issuing and Distribution Accounts section](../../../tokens/control-asset-access.mdx#issuing-and-distribution-accounts):

```rust
#![cfg(test)]

use soroban_sdk::testutils::Address as _;
use soroban_sdk::{token, Address, Env};
use token::TokenClient;
use token::StellarAssetClient;

#[test]
fn test() {
    let e = Env::default();
    e.mock_all_auths();

    let issuer = Address::random();
    let distributer = Address::random();

    let token_address = e.register_stellar_asset_contract(issuer.clone());

    // client for SEP-41 operations
    let token = TokenClient::new(&e, &token_address);

>     // client for Stellar Asset Contract operations

Reviewer: I think using the word `functions` instead of `operations` _might_ make more sense in these explainer comments. I can see how, when we're using SAC functions, it's analogous to the "classic" tx operations, so it makes sense to me. But, if someone is unfamiliar with those ops, it might not.

What I'm **sure of** is that it's splitting hairs. I would absolutely not die on this hill 🤣

Author: Makes sense, here we are in a dev context so functions seems better 👍

    let token_sac = StellarAssetClient::new(&e, &token_address);

>     // 1e7 diff between minimal unit and unit itself

Reviewer: I understand what this means, but the wording feels a little "clunky"? Can we re-phrase this so it's a little clearer?

Author: oh yes, I will rephrase.

    let genesis_amount: i128 = 1_000_000_000 * 10_000_000;

    token_sac.mint(&issuer, &distributer, &genesis_amount);

    assert_eq!(token.balance(&distributor), genesis_amount);
```

### Examples

See the full examples that utilize the token contract in various ways for more details:

> - [Timelock](../../smart-contracts/example-contracts/timelock.mdx) and [single offer](../../smart-contracts/example-contracts/single-offer-sale.mdx) move token via `xfer` to and from the contract

Reviewer: ```suggestion
- [Timelock](../../smart-contracts/example-contracts/timelock.mdx) and [single offer](../../smart-contracts/example-contracts/single-offer-sale.mdx) move token via `transfer` to and from the contract
```

> - [Atomic swap](../../smart-contracts/example-contracts/atomic-swap.mdx) uses `incr_allow` to transfer token on behalf of the user

Reviewer: This example no longer uses `allowance` functions, but does use the `transfer` function. The text on that page looks like it's pretty out-dated and will need to be updated sometime soon.

I quickly looked through the example contracts, and it looks like there isn't really any use of the `approve` or `allowance` functions, outside of the `token` example and a test or two.

Author: For now I said it's `transfer` too and changed the references to `incr_allow` as not used in the example.
